pkg/prometheus/common.go

```diff
@@ -594,3 +598,31 @@ func MakeProbes(
 
 	return startupProbe, readinessProbe, livenessProbe
 }
+
+// If a ServiceName is specified, check that the service exists in the namespace. If it does not exist, fail
+// the reconciliation.
+// Also, ensure that the Prometheus instance is selected by the service. If not, fail the reconciliation.
+func CheckCustomService(ctx context.Context, serviceName *string, prometheusNS string, prometheusName string, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
+	svc, err := svcClient.Get(ctx, *serviceName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Errorf("service %s/%s does not exist: %w", prometheusNS, *serviceName, err)
+		}
+
+		return err
+	}
+
+	// Get the user defined label selectors from the service.
+	svcSelector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: svc.Spec.Selector})
+	if err != nil {
+		return fmt.Errorf("failed to get label selector while synchronizing service: %w", err)
+	}
+
+	// Check if the svcSelector is the same as the Selector used to match the pods handled by this
+	// Prometheus resource.
+	if !svcSelector.Matches(labels.Set(selectorLabels)) {
+		return fmt.Errorf("service %s/%s does not select prometheus server %s", prometheusNS, *serviceName, prometheusName)
```

Comment:
(suggestion) let's print out the details.
```suggestion
		return fmt.Errorf("service %s/%s  with selector %q does not select Prometheus pods %s with labels %q", prometheusNS, *serviceName, prometheusName, svcSelector.String(), labels.Set(selectorLabels).String())
```

Reply:
Changed it a bit
```
		return fmt.Errorf("service %s/%s with selector %s does not select Prometheus sset pods %s/%s with labels %s",
			prometheusNS, serviceName, svcSelector.String(), prometheusNS, prometheusName, labels.Set(selectorLabels).String())
```